File: backend/app/schemas/nearby_bank.py
"""Schemas for real bank branches discovered from OpenStreetMap.

These are *not* channel partners. A discovered bank is a public map feature:
we know where it is and what it is called, and nothing about its operational
health. It therefore carries no quota, no NPA, and no Partner Health Score,
and it can never become an application's `partner_id`.
"""

from pydantic import BaseModel, ConfigDict, Field


class NearbyBank(BaseModel):
    """One real bank branch from OpenStreetMap.

    `osm_id` is the stable OSM identifier in `<type>/<id>` form (for example
    `node/1234567`), which is what OSM itself uses to address a feature and
    keeps node and way ids from colliding.
    """

    model_config = ConfigDict(extra="forbid")

    osm_id: str
    name: str
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)
    distance_km: float = Field(ge=0)

    # Absent whenever OSM has no address tags for the feature. Most Indian
    # bank nodes carry none, so this is null far more often than not and is
    # never synthesised from the coordinates.
    address: str | None = None


class NearbyBankResponse(BaseModel):
    """Discovery result, ordered nearest-first.

    `source` names the upstream so the client can attribute the data, and
    `cached` says whether this answer avoided an upstream call.

    `total` counts what is in `items`, while `discovered` counts everything
    found inside the radius. They differ only when `capped` is true, and the
    pair exists so the screen can say "the 50 nearest of 1,041" rather than
    presenting a truncated list as if it were the whole neighbourhood.
    """

    model_config = ConfigDict(extra="forbid")

    items: list[NearbyBank]
    total: int = Field(ge=0)
    discovered: int = Field(ge=0)
    capped: bool = False
    source: str = "openstreetmap"
    cached: bool = False

File: frontend/lib/models/nearby_bank.dart
/// A real bank branch discovered from OpenStreetMap.
///
/// Deliberately *not* a [RecommendedPartner]. A discovered bank is a public
/// map feature: we know where it is and what it is called, and nothing about
/// its operational health. It therefore has no quota, no NPA, and no Partner
/// Score, and it can never be sent as an application's `partner_id` — the
/// backend would reject the id anyway, since it belongs to OpenStreetMap
/// rather than to our channel-partner table.
class NearbyBank {
  const NearbyBank({
    required this.osmId,
    required this.name,
    required this.latitude,
    required this.longitude,
    required this.distanceKm,
    this.address,
  });

  /// Stable OpenStreetMap identifier in `<type>/<id>` form, e.g. `node/123`.
  /// Used as the widget key and for marker/card correspondence.
  final String osmId;

  final String name;
  final double latitude;
  final double longitude;

  /// Great-circle distance from the applicant's stored coordinates, in
  /// kilometres, measured with the same Haversine formula the partner
  /// router uses.
  final double distanceKm;

  /// Only present when OpenStreetMap genuinely has address tags for the
  /// feature, which is the minority of Indian bank nodes. Never derived from
  /// the coordinates.
  final String? address;

  factory NearbyBank.fromJson(Map<String, dynamic> json) {
    double toDouble(Object? value) =>
        value is num ? value.toDouble() : double.parse(value.toString());

    final address = json['address'];
    return NearbyBank(
      osmId: json['osm_id'].toString(),
      name: json['name'].toString(),
      latitude: toDouble(json['latitude']),
      longitude: toDouble(json['longitude']),
      distanceKm: toDouble(json['distance_km']),
      // Treat an empty string as absent so the UI never renders a blank line.
      address: (address == null || address.toString().trim().isEmpty)
          ? null
          : address.toString(),
    );
  }
}
